src/main/java/org/openstreetmap/atlas/geography/boundary/DynamicGridIndexBuilder.java

```diff
@@ -1,40 +1,223 @@
+            final Polygon polygon = item.polygon;
+            Envelope box = null;
+
+            try
+            {
+                box = new Envelope(minX, maxX, minY, maxY);
+                final double width = maxX - minX;
+                final double height = maxY - minY;
+                final Polygon geoBox = buildGeoBox(minX, maxX, minY, maxY);
+
+                if (!geoBox.intersects(polygon))
+                {
+                    // We throw away non-intersected cells
+                    return;
+                }
+                else if (Math.max(width / 2, height / 2) < GRANULARITY)
+                {
+                    synchronized (this.index)
+                    {
+                        this.index.insert(box, polygon);
+                    }
+                    return;
+                }
+                else if (CountryBoundaryMap.isSameCountry(this.rawIndex.query(box))
+                        || polygon.covers(geoBox))
+                {
+                    // The box must be within the boundary envelope so rawIndex query should return
+                    // at least the boundary itself. If query results isSameCountry returns true,
+                    // the box must ONLY intersect with the country the boundary belongs, in this
+                    // case we treat it same as inside cell. This is an end point, we add the box to
+                    // the spatial index.
+                    synchronized (this.index)
+                    {
+                        this.index.insert(box, polygon);
+                    }
+                    return;
+                }
+
+                // None of above conditions are met, we need to split the box into sub-boxes
+                if (width > height)
+                {
+                    this.queue.add(new GridWorkItem(minX, minY, minX + width / 2, maxY, polygon));
+                    this.queue.add(new GridWorkItem(minX + width / 2, minY, maxX, maxY, polygon));
+                }
+                else
+                {
+                    this.queue.add(new GridWorkItem(minX, minY, maxX, minY + height / 2, polygon));
+                    this.queue.add(new GridWorkItem(minX, minY + height / 2, maxX, maxY, polygon));
+                }
+            }
+            catch (final TopologyException e)
+            {
+                // Catch mostly JTS Topology exceptions
+                final String countryCode = CountryBoundaryMap.getGeometryProperty(polygon,
+                        ISOCountryTag.KEY);
+                logger.error("Unable to build tree under box {} for country code {}.", box,
+                        countryCode, e);
+                // Save the current box and do not iterate down.
+                synchronized (this.index)
+                {
+                    this.index.insert(box, polygon);
+                }
+            }
+            catch (final Exception e)
+            {
+                logger.error("Processor failed.", e);
```

Comment:
What's the impact of failure here? Corrupted grid index?

Reply:
Yes, possibly corrupted index. I'll let it throw here.